onnx/defs/nn/defs.cc

```diff
@@ -2832,4 +2833,392 @@ ONNX_OPERATOR_SET_SCHEMA(
+              .Add("QSeqLen = Shape <start = -2, end = -1> (Q)") // q_sequence_length
+              .Add("KVSeqLen = Shape <start = 2, end = -1> (K)") // kv_sequence_length
+              .Const1D("NegOne", static_cast<int64_t>(-1)) // head_size, inferred from other dimensions
+              .Add("QNewShape = Concat <axis = 0> (BatchSize, QNumHeads, QSeqLen, NegOne)")
+              .Add("KVNewShape = Concat <axis = 0> (BatchSize, KVNumHeads, KVSeqLen, NegOne)")
+              .Add("QReshaped = Reshape (Q, QNewShape)")
+              .Add("KReshaped = Reshape (K, KVNewShape)")
+              .Add("VReshaped = Reshape (V, KVNewShape)");
+
+          // Calculate scaling factor if scale attribute not provided
+          auto scale_attr = ctx.getAttribute("scale");
+          float scale = (scale_attr != nullptr) ? scale_attr->f() : static_cast<float>(1);
+          builder
+              .Add("QKHeadSize = Shape <start = 3, end = 4> (QReshaped)") // head_size for Q and K
+              .Add("SqrtHeadSize = Sqrt(QKHeadSize)")
+              .Const1D("One1D", static_cast<int64_t>(1))
+              .Const1D("Zero1D", static_cast<int64_t>(0))
+              .Add("CalculatedScale = Div(One1D, SqrtHeadSize)")
+              .Const("ScaleF", ToTensor<float>(scale))
+              .Add(scale_attr != nullptr ? "ScaleFactor = Identity(ScaleF)" : "ScaleFactor = Identity(CalculatedScale)")
+              .Add("ScaleFactorF = Cast (ScaleFactor)", "to", T);
+
+          // Update key and value caches for past and present states
+          if (ctx.hasOutput(1)) {
+            if (ctx.hasInput(4)) {
+              builder.Add("present_key = Concat <axis = 2> (past_key, KReshaped)");
+            }
+          } else {
+            builder.Add("present_key = Identity (KReshaped)");
+          }
+
+          if (ctx.hasOutput(2)) {
+            if (ctx.hasInput(5)) {
+              builder.Add("present_value = Concat <axis = 2> (past_value, VReshaped)");
+            }
+          } else {
+            builder.Add("present_value = Identity (VReshaped)");
+          }
+
+          // Create a attn_bias filled with zeros of shape (q_sequence_length, kv_sequence_length)
+          builder.Add("NewKVSeqLen =  Shape <start = -2, end = -1> (present_key)")
+              .Add("AttnBiasShape = Concat <axis = -1> (QSeqLen, NewKVSeqLen)")
+              .Add("AttnBiasZeros = ConstantOfShape(AttnBiasShape)");
+
+          // If attn_mask is provided
+          if (ctx.hasInput(3)) {
+            auto* up = ctx.getInputType(3);
+            if ((up == nullptr) || (!up->has_tensor_type()))
+              return false;
+            int64_t U = up->tensor_type().elem_type();
+            builder
+                .Add(
+                    U == ONNX_NAMESPACE::TensorProto_DataType_BOOL
+                        ? "MaskedAttnBias = Where(attn_mask, AttnBiasZeros, FloatInf)"
+                        : "MaskedAttnBias = Add(attn_mask, AttnBiasZeros)")
```

Comment:
Why add zero here? Is this doing any broadcast-reshaping? Or, would "Identity(attn_mask)" work here?